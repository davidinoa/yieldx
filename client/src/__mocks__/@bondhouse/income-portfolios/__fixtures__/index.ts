import {
  OrderDirectionEnum,
  PortfolioStatus,
  Proposal
} from "@bondhouse/income-portfolios";

const now = new Date().toISOString();
export const PROPOSAL_MOCK: Proposal = {
  proposalId: "mockProposalId",
  updatedAt: now,
  commentary: [
    {
      topic: "Income",
      content:
        "The expected monthly income is $1008 for an annualized yield of 6.05%."
    },
    {
      topic: "Risk",
      content:
        "You have a 29% chance of losing money in a given year, and the portfolio's largest drop in account value was 14.2% in the last five years. This is -4.4% less risk than similar securities yielding 6-7%."
    },
    {
      topic: "Expenses",
      content:
        "The weighted expense ratio is 0.41% per year. This is 1.12% less than similar funds with this yield."
    }
  ],
  portfolio: {
    status: PortfolioStatus.READY,
    id: "e4d4c1be-ad5b-4221-8e34-12e0e8d73dcf",
    currency: "USD",
    name: "Your new portfolio",
    positions: [],
    createdAt: now,
    updatedAt: now,
    metadata: { app: "inpaas" }
  },
  incomePortfoliosAnalysis: {
    income: 1007.5426586045,
    risk: 9.785923627675485,
    expenseRatio: 0.41044404884999997,
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    yield: 6.045255951627,
    afterTaxYield: 3.6271535709762004,
    maxDrawdown: -14.161982640325931
  },
  proposalRequest: {
    goal: {
      income: 600
    },
    investmentAmount: 60_000
  },
  orders: [
    {
      assetId: "102dd159-1401-49dc-bad0-9cb776fbe5cb",
      direction: OrderDirectionEnum.BUY,
      quantity: 1209
    },
    {
      assetId: "6f5c3533-89be-48ab-806a-ab693cf0a7ff",
      direction: OrderDirectionEnum.BUY,
      quantity: 933.9999999999999
    },
    {
      assetId: "8feba4a4-1a13-47be-b447-1752287b8e2f",
      direction: OrderDirectionEnum.BUY,
      quantity: 398
    },
    {
      assetId: "4357f998-6755-40e6-8607-adbd454f9756",
      direction: OrderDirectionEnum.BUY,
      quantity: 217
    },
    {
      assetId: "a0135762-04b7-4dc6-8077-3ef61bcedeaa",
      direction: OrderDirectionEnum.BUY,
      quantity: 114
    },
    {
      assetId: "f10c6a1d-c607-4666-8621-8b1a2c3741b3",
      direction: OrderDirectionEnum.BUY,
      quantity: 489
    },
    {
      assetId: "e4871590-916e-4256-b957-18bb2e8e8b2f",
      direction: OrderDirectionEnum.BUY,
      quantity: 295.0000096246391
    },
    {
      assetId: "621e7f48-e332-4e92-a875-260ed4ec9569",
      direction: OrderDirectionEnum.BUY,
      quantity: 2787
    },
    {
      assetId: "ac1ab82c-c1c4-4d34-83fa-df78c79404bf",
      direction: OrderDirectionEnum.BUY,
      quantity: 714
    },
    {
      assetId: "USD",
      direction: OrderDirectionEnum.SELL,
      quantity: -199999.99999999997
    }
  ]
};
