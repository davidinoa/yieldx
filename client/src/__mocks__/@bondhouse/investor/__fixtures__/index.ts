import { AccountType, LinkedBankAccountStatus,
  Amount,
  EmploymentStatus, InvestmentExperience, InvestmentObjective,
  InvestorType,
  LiquidityNeeds,
  PhoneType, RiskTolerance,
  TimeHorizon,
  TrustedContact, Worth} from "@bondhouse/investor";

export const TD_SAVINGS_ACTIVE = {
  "id": "781359af-35fe-4399-8351-cb93f239d484",
  "investorId": "b7ad4d69-e732-4bec-ba31-a256b22a51e4",
  "mask": "1111",
  "bankName": "TD Bank",
  "accountType": AccountType.SAVINGS,
  "status": LinkedBankAccountStatus.ACTIVE,
  "createdAt": "2020-06-18T20:03:58.585590Z",
  "updatedAt": "2020-06-18T20:03:58.585590Z"
};

export const INDIVIDUAL_INVESTOR = {
  id: "investorId",
  investorType: InvestorType.INDIVIDUAL,
  investorInfo: {
    givenName: "Alfredo",
    familyName: "Alpizar",
    citizenshipCountry: "USA",
    contactInformation: {
      streetAddress: "372 Westminster Road",
      city: "Brooklyn",
      state: "NY",
      postalCode: "11218",
      country: "USA",
      phone: "347-123-4567",
      phoneType: PhoneType.MOBILE,
      emailAddresses: ["test@testing.com"]
    },
    trustedContactInfo: { trustedContact: TrustedContact.EXCLUDE },
    disclosures: {
      isControlPerson: false,
      isAffiliatedExchangeOrFINRA: false,
      isPoliticallyExposed: false
    },
    suitabilityProfile: {
      timeHorizon: TimeHorizon.AVERAGE,
      liquidityNeeds: LiquidityNeeds.VERYIMPORTANT
    },
    applicantSignature: { eSigned: true },
    birthCountry: "USA",
    employment: { employmentStatus: EmploymentStatus.STUDENT },
    investmentProfile: {
      investmentObjective: InvestmentObjective.INCOME,
      investmentExperience: InvestmentExperience.LIMITED,
      annualIncomeUSD: Amount.FROM100TO200K,
      liquidNetWorthUSD: Worth.UPTO50K,
      totalNetWorthUSD: Worth.UPTO50K,
      riskTolerance: RiskTolerance.MEDIUM,
      federalTaxBracketPercent: 5
    }
  },
  createdAt: "2020-05-18T15:06:53.540735Z",
  updatedAt: "2020-05-18T15:06:53.540735Z"
};
