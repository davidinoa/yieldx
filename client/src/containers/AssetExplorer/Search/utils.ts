interface InGridFilter {
  filter: string | number;
  filterTo: number;
  filterType: string;
  type: string;
  dateFrom: string;
  dateTo: string;
  values: string[];
}

export interface AgGridFilters {
  [key: string]: any;
}

const getFilter = (filter: InGridFilter) => {
  /**
   *  The filter object from Ag Grid API. In the case of filtering `name` column.
   *  { filter: "Asset Name", filterType: "text", type: "contains"}
   */
  let searchFilter = {};
  const {
    filter: filterContent,
    filterType,
    type,
    filterTo,
    values,
    dateFrom,
    dateTo
  } = filter;
  if (filterType === "text") {
    switch (type) {
      case "contains":
        searchFilter = { containsText: filterContent };
        break;
      case "notContains":
        searchFilter = { doesNotContainText: filterContent };
        break;
      case "equals":
        searchFilter = { equalsText: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualText: filterContent };
        break;
      case "startsWith":
        searchFilter = { startsWithText: filterContent };
        break;
      case "endsWith":
        searchFilter = { endsWith: filterContent };
        break;
      default:
        searchFilter = { containsText: filterContent };
        break;
    }
  } else if (filterType === "number") {
    switch (type) {
      case "equals":
        searchFilter = { equalsNumber: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualNumber: filterContent };
        break;
      case "lessThan":
        searchFilter = { isLessThanNumber: filterContent };
        break;
      case "lessThanOrEqual":
        searchFilter = { isLessThanOrEqualToNumber: filterContent };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanNumber: filterContent };
        break;
      case "greaterThanOrEqual":
        searchFilter = { isGreaterThanOrEqualToNumber: filterContent };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToNumber: filterContent,
          isLessThanOrEqualToNumber: filterTo
        };
        break;
      default:
        searchFilter = { equalsNumber: filterContent };
        break;
    }
  } else if (filterType === "date") {
    const convertedDateFrom = new Date(dateFrom);
    const convertedDateTo = new Date(dateTo);
    switch (type) {
      case "equals":
        searchFilter = { equalsDate: convertedDateFrom };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanDate: convertedDateFrom };
        break;
      case "lessThan":
        searchFilter = { isLessThanDate: convertedDateFrom };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualDate: convertedDateFrom };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToDate: convertedDateFrom,
          isLessThanOrEqualToDate: convertedDateTo
        };
        break;
      default:
        searchFilter = { equalsDate: convertedDateFrom };
        break;
    }
  } else if (filterType === "set" && values) {
    // We may need to add `set` type Ag grid column filter params like cusip or isin
    // in the search request in the future.
    // searchFilter = { inText: values };
  }
  return searchFilter;
};

export const convertInGridFilters = (filters: AgGridFilters) => {
  const convertedFilters: { [index: string]: any } = {};
  Object.keys(filters).forEach(key => {
    const inGridFilter = filters[key];
    if (key === "minimumQuantity") key = "bond.minimumDenomination";
    convertedFilters[key] = getFilter(inGridFilter);
  });

  return convertedFilters;
};
