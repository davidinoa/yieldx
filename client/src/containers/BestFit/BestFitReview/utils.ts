import { getRatingValue } from "../../../util/ratings";
import YieldXPortfolio, {
  YieldXPosition
} from "../../../models/YieldXPortfolio";

export function getBondRatingsData(
  portfolio: YieldXPortfolio
): { name: string; value: number }[] {
  if (portfolio.getBondSleeve) {
    const ratingMarketValues: { [key: string]: number } = {};
    const bondPortfolio = portfolio.getBondSleeve();
    bondPortfolio.positions.forEach((position: YieldXPosition) => {
      const { asset, weight } = position;
      const rating = asset.rating ?? "NR";
      ratingMarketValues[rating] = (ratingMarketValues[rating] || 0) + weight;
    });
    return Object.entries(ratingMarketValues)
      .map(([rating, weight]) => {
        return {
          name: rating,
          value: weight
        };
      })
      .sort((a, b) => {
        return getRatingValue(a.name) - getRatingValue(b.name);
      });
  }
  return [];
}
