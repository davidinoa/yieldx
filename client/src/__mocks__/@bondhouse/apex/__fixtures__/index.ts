import { AccountStatus } from "@bondhouse/apex";

export const MOCK_APEX_ACCOUNT = {
  id: "f42bebf5-c951-4b55-a709-b2339aaf3e94",
  investorId: "investorId",
  apexRequestId: "157bec3c-323b-41d9-b2a0-368e6493f48c",
  apexAccountNumber: "7XL08686",
  status: AccountStatus.ACTIVE,
  createdAt: "2020-05-18T15:06:55.121931Z",
  updatedAt: "2020-05-26T22:56:56.835664Z"
};
