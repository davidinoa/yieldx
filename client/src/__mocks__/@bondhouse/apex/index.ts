import { AccountStatus } from "@bondhouse/apex";
import { MOCK_APEX_ACCOUNT } from "./__fixtures__";
// eslint-disable-next-line @typescript-eslint/no-empty-function,no-shadow
function ApexAccountsApi() {}

ApexAccountsApi.prototype.getApexAccountByInvestorId = jest.fn(() =>
  Promise.resolve({
    data: {
      apexAccount: MOCK_APEX_ACCOUNT
    }
  })
);

// eslint-disable-next-line @typescript-eslint/no-empty-function,no-shadow
function StatementsApi() {}

StatementsApi.prototype.getAccountStatements = jest.fn(() =>
  Promise.resolve([
    {
      apexAccountId: "28b5b427-053b-4ad8-9acb-05a5e7bbe614",
      statmentDate: "12/31/2014",
      documentType: "Account Statement",
      statementURL:
        "https://UATEDocuments.apexclearing.com/RedirectConfirmURL.aspx?ds=1DC69CF82426F431D4535F6549270DAC58EEB05D61A8209E49CAAE3BC76EE8B866CEDC5A04A68284717045A7352477CC49D5E186BA7CA707C0EF8A692C654BED5732532662D182181AAF4E3C7F5B219956BF0F1B6F57C588C0465DF32D3A5EB58EFB4CFF188B33275E8BFB631B97C80DFEB0838AECFC00DF537E96EB28868EFF10660E83DD1A6C5808E6DED036C15E25336F448E9F69DF30118E6960DD1F98D4B8E4714BCAE3D1435A79505BC3FB730B09FEA86B8E3DEB1F8DDAF65E746FBE833308839C1FA0381057F4E71EFC7E14AF5DBB327B2CA848EC4CC2D036AF66C648E78AD70688C550DEB37F942CC82AEEE4F6E1BA95C082DB8D326BC86D5B5DC113060E7449A97FD2F019A66050CA0AF83B0FA50FFCFE9F5E1664C92DA902988BDF&pid=TkVXU1RNVHx2NzEyNi01Mj9yLTUyOTMtNzc1xS1PREEyLTI4MzBGQUFZLTM5NzAxLTU0MzIxNC00MjE4NzItMjE5MjAyLTg1LTc5LTI2NDMtODktMC1eATAxNAE1UEkBMzIyNzIBTUFSSyBTTEFWT05JQQExN1EwNgFXRUEBAgECAQIBAgExNDQBMTcxMDgBMTMxODQBAgECAQIBA0101816"
    }
  ])
);

export { ApexAccountsApi, AccountStatus, StatementsApi };
