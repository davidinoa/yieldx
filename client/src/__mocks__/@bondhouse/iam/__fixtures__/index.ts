import { User } from "@bondhouse/iam";
import { UserType } from "..";

export const mockAdminUser: User = {
  id: "id-admin",
  email: "mock.admin@bondhouse.com",
  first: "Mock",
  last: "Admin",
  userType: UserType.ADMIN,
  terms: true
};
