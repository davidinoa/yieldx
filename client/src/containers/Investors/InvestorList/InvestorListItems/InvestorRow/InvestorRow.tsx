import React from "react";
import dayjs from "dayjs";
import { InvestorWithApexAccount } from "@bondhouse/investor";
import { AccountStatus } from "@bondhouse/apex";
import {
  CTAColumn,
  DetailsButton,
  ListRow,
  Status,
  StatusColumn
} from "components/ListPages/ListPages.styles";
import { List } from "components/ListPages/List";

interface Props {
  account: InvestorWithApexAccount;
}

export default function InvestorRow({ account }: Props) {
  const { apexAccount, investor } = account;
  const { id, investorInfo, investorType, createdAt } = investor;
  const accountCreatedInLastMinute = dayjs()
    .subtract(2, "minute")
    .isBefore(dayjs(createdAt));
  const { familyName, givenName } = investorInfo;
  const { status } = apexAccount || {
    status: AccountStatus.ERROR
  };
  return (
    <ListRow key={id}>
      <StatusColumn>
        {apexAccount === undefined && !accountCreatedInLastMinute && (
          <Status className="error" />
        )}
        {accountCreatedInLastMinute && <Status className="pending" />}
        {apexAccount !== undefined && (
          <Status className={status.toLocaleLowerCase()} />
        )}
      </StatusColumn>
      <List.Col
        className="p-0"
        label="INVESTOR NAME"
        value={`${givenName} ${familyName}`}
      />
      <List.Col
        label="TYPE"
        value={investorType}
        className="col-4 col-xxl-3 centered"
      />
      <List.Col
        label="CREATED AT"
        className="col-4 col-xxl-3 centered"
        value={dayjs(createdAt).format("MMM D, YYYY")}
      />

      <CTAColumn>
        <DetailsButton
          data-testid="InvestorList-ShowInvestorLink"
          data-disabled={false}
          to={`/investors/${id}`}
        >
          Details
        </DetailsButton>
      </CTAColumn>
    </ListRow>
  );
}
