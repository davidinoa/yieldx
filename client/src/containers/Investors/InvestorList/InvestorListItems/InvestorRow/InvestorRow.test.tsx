import React from "react";
import { render, screen } from "test/utils";
import InvestorRow from "./InvestorRow";
import { INDIVIDUAL_INVESTOR } from "__mocks__/@bondhouse/investor/__fixtures__";
import { MOCK_APEX_ACCOUNT } from "__mocks__/@bondhouse/apex/__fixtures__";

describe("InvestorRow", () => {
  test("Should render", async () => {
    const { container } = render(
      <InvestorRow
        account={{
          investor: INDIVIDUAL_INVESTOR,
          apexAccount: MOCK_APEX_ACCOUNT
        }}
      />
    );
    expect(container).toBeInTheDocument();
  });

  test("Should display required columns", () => {
    render(
      <InvestorRow
        account={{
          investor: INDIVIDUAL_INVESTOR,
          apexAccount: MOCK_APEX_ACCOUNT
        }}
      />
    );
    expect(screen.getByText(/INVESTOR NAME/i)).toBeInTheDocument();
    expect(screen.getByText(/TYPE/i)).toBeInTheDocument();
    expect(screen.getByText(/CREATED AT/i)).toBeInTheDocument();
    expect(
      screen.getByTestId("InvestorList-ShowInvestorLink")
    ).toBeInTheDocument();
  });

  test("Status column should render appropriately while fetching apex account status", async () => {
    const { asFragment } = render(
      <InvestorRow
        account={{
          investor: INDIVIDUAL_INVESTOR,
          apexAccount: MOCK_APEX_ACCOUNT
        }}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
