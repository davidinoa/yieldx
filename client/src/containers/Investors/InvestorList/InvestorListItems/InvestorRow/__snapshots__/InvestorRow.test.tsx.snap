// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvestorRow Status column should render appropriately while fetching apex account status 1`] = `
<DocumentFragment>
  .c3 {
  font-weight: bold;
  color: var(--text);
  font-size: 0.725rem;
  margin: 0;
  line-height: normal;
}

.c4 {
  margin: 0;
  font-weight: bold;
  font-size: 0.6rem;
  -webkit-letter-spacing: 0.2em;
  -moz-letter-spacing: 0.2em;
  -ms-letter-spacing: 0.2em;
  letter-spacing: 0.2em;
  text-transform: uppercase;
  color: var(--subtext-3);
}

.c0 {
  height: 70px;
  list-style: none;
  border: 1px solid var(--light);
  background: var(--dark);
  border-radius: 5px;
  position: relative;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  margin-bottom: 0.45rem;
}

.c0 .no-borders::after {
  border-right: none !important;
  border-left: none !important;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 5px;
  width: 5px;
  border-radius: 50%;
  margin: 0 auto;
}

.c2.active,
.c2.ready {
  background: var(--active);
}

.c2.deactivated,
.c2.inactive,
.c2.blocked,
.c2.closed,
.c2.terminated {
  background: var(--subtext-3);
}

.c2.tracking {
  background: var(--tracking);
}

.c2.pending {
  background: var(--caution);
}

.c2.error {
  background: var(--issue);
}

.c5 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: var(--page);
  font-weight: bold;
  border-radius: 0.25rem;
  color: white;
  -webkit-text-decoration: none;
  text-decoration: none;
  width: 70px;
  height: 26px;
  font-size: 0.8rem;
}

.c5:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: white;
}

@media (min-width:1600px) {
  .c3 {
    font-size: 0.875rem;
  }
}

@media (min-width:1600px) {
  .c4 {
    font-size: 0.6875rem;
  }
}

@media (min-width:1600px) {
  .c0 {
    padding-top: 0.75rem;
    padding-bottom: 0.75rem;
    margin-bottom: 0.575rem;
  }
}

@media (min-width:1600px) {
  .c2 {
    height: 6px;
    width: 6px;
  }
}

@media (min-width:1600px) {
  .c5 {
    width: 90px;
    height: 30px;
    font-size: 0.975rem;
  }
}

<div
    class="c0 row"
  >
    <div
      class="c1 col-1 no-borders"
    >
      <i
        class="c2 active"
      />
    </div>
    <div
      class="d-flex flex-column justify-content-center p-0 col"
    >
      <span
        class="c3"
      >
        Alfredo Alpizar
      </span>
      <span
        class="c4"
      >
        INVESTOR NAME
      </span>
    </div>
    <div
      class="d-flex flex-column justify-content-center col-4 col-xxl-3 centered col"
    >
      <span
        class="c3"
      >
        INDIVIDUAL
      </span>
      <span
        class="c4"
      >
        TYPE
      </span>
    </div>
    <div
      class="d-flex flex-column justify-content-center col-4 col-xxl-3 centered col"
    >
      <span
        class="c3"
      >
        May 18, 2020
      </span>
      <span
        class="c4"
      >
        CREATED AT
      </span>
    </div>
    <div
      class="d-flex flex-column col-3 col-xxl-2 centered justify-content-center"
    >
      <a
        class="c5"
        data-disabled="false"
        data-testid="InvestorList-ShowInvestorLink"
        href="/investors/investorId"
      >
        Details
      </a>
    </div>
  </div>
</DocumentFragment>
`;
